include/eosio/vm/host_function.hpp

```diff
@@ -17,555 +20,353 @@
+            return preconditions_runner<I+1, Preconditions>(ctx,
+                  precondition_runner<std::tuple_element_t<I, Preconditions>>(ctx, std::move(tup), std::make_index_sequence<sizeof...(Args)>{}));
          else
-            return detail::resolve_result(detail::init_wasm_type_converter(wasm_type_converter<T>{}, alloc)
-                                          .to_wasm(static_cast<T&&>(res)), alloc);
+            return static_cast<std::tuple<Args...>&&>(tup);
       }
    } //ns detail
 
-   template <>
-   struct wasm_type_converter<bool> {
-      static bool from_wasm(uint32_t val) { return val != 0; }
-      static uint32_t to_wasm(bool val) { return val? 1 : 0; }
-   };
-
-   template<typename T, std::size_t Align>
-   struct wasm_type_converter<aligned_ptr<T, Align>> {
-      aligned_ptr_wrapper<T, Align> from_wasm(void* ptr) { return {ptr}; }
-   };
-
-   template<typename T, std::size_t Align>
-   struct wasm_type_converter<aligned_ref<T, Align>> {
-      aligned_ptr_wrapper<T, Align> from_wasm(void* ptr) { return {ptr}; }
-   };
-
-   template <typename T>
-   inline constexpr auto to_wasm_type() -> uint8_t {
-      if constexpr (std::is_same_v<T, void>)
-         return types::ret_void;
-      else if constexpr (std::is_same_v<T, bool>)
-         return types::i32;
-      else if constexpr (std::is_integral_v<T> && sizeof(T) == 4)
-         return types::i32;
-      else if constexpr (std::is_integral_v<T> && sizeof(T) == 8)
-         return types::i64;
-      else if constexpr (std::is_floating_point_v<T> && sizeof(T) == 4)
-         return types::f32;
-      else if constexpr (std::is_floating_point_v<T> && sizeof(T) == 8)
-         return types::f64;
-      else if constexpr (std::is_pointer_v<T> || std::is_reference_v<T>)
-         return types::i32;
-      else
-         return vm::to_wasm_type<typename detail::get_wasm_type<T>::type>();
+   template <typename... Args, typename F>
+   void invoke_on(F&& func, Args&&... args) {
+      detail::invoke_on_impl(static_cast<F&&>(func), std::forward<Args>(args)...);
    }
 
-   template <uint8_t Type>
-   struct _to_wasm_t;
-
-   template <>
-   struct _to_wasm_t<types::i32> {
-      typedef i32_const_t type;
-   };
+#define EOS_VM_INVOKE_ON(CONDITION) \
+    invoke_on(CONDITION, std::forward<Args>(args)...);
 
-   template <>
-   struct _to_wasm_t<types::i64> {
-      typedef i64_const_t type;
+#define EOS_VM_PRECONDITION(NAME, ...)                                       \
+   struct NAME {                                                             \
+      template <typename Type_Converter, typename... Args>                   \
+      inline static decltype(auto) condition(Type_Converter& ctx, Args&&... args) { \
+        __VA_ARGS__;                                                         \
+        return std::forward_as_tuple(args...);                               \
+      }                                                                      \
```

Comment:
Once again, eyebleed.

Reply:
The only part that the user defines is a lambda.  So, just allow registered_host_functions::add to take a lambda instead of this whole complicated mess.